myApp/src/app/app.component.html

```diff
@@ -0,0 +1,30 @@
+<form [formGroup]="myForm" (ngSubmit)="onSubmit()" class="form">
+  <div class="field">
+    <label for="username">username</label>
+    <input type="text" id="username" formControlName="username" class="input">
+  </div>
+  <div *ngIf="myForm.get('username')?.invalid && myForm.get('username')?.touched" class="error">
+    <p *ngIf="myForm.get('username')?.hasError('required')" class="invalid">Username is required</p>
+    <p *ngIf="myForm.get('username')?.hasError('minlength')" class="invalid">2 characters minimum</p>
+    <p *ngIf="myForm.get('username')?.hasError('maxlength')" class="invalid">20 characters maximum</p>
+  </div>
+  <div class="field">
+    <label for="email">email</label>
+    <input type="text" id="email" formControlName="email" class="input">
+  </div>
+  <div *ngIf="myForm.get('email')?.invalid && myForm.get('email')?.touched" class="error">
+    <p *ngIf="myForm.get('email')?.hasError('required')" class="invalid">Email is required</p>
+    <p *ngIf="myForm.get('email')?.hasError('email')" class="invalid">not valid email</p>
+  </div>
+  <div class="field">
+    <label for="password">password</label>
+    <input type="password" id="password" formControlName="password" class="input">
+  </div>
+  <div *ngIf="myForm.get('password')?.invalid && myForm.get('password')?.touched" class="error">
+    <p *ngIf="myForm.get('password')?.hasError('required')" class="invalid">Password is required</p>
+    <p *ngIf="myForm.get('password')?.hasError('minlength')" class="invalid">10 characters minimum</p>
+  </div>
+  <div class="button">
+    <button type="submit" class="button">Register</button>
+  </div>
+</form>
```

Comment:
This is very repetative and hard to maintain. you see here you will need to add a <p> and an ngif for each error? I would like to be able to add validators to the array with a message that is all in one place. 
The error messages are all over the place. the HTML needs to be looked at very carefully when adding validations. 
Think how to extract the logic and make it easily maintainable.